Overhaul importmulti logic

sipa opened the pull request:
This is an alternative to #14558 (it will warn when fields are being ignored). In addition:
* It makes sure no changes to the wallet are made when an error in the input exists.
* It validates all arguments, and will fail if anything fails to parse.
* Adds a whole bunch of sanity checks

sipa committed:
Overhaul importmulti logic

This introduces various changes to the importmulti logic:
* Instead of processing input and importing things at the same time, first
  process all input data and verify it, so no changes are made in case of
  an error.
* Verify that no superfluous information is provided (no keys or scripts
  that don't contribute to solvability in particular).
* Add way more sanity checks, by means of descending into all involved
  scripts.

fanquake commented:
Travis is sad about trailing whitespace:
```
This diff appears to have added new lines with trailing whitespace.
The following changes were suspected:
diff --git a/src/wallet/rpcdump.cpp b/src/wallet/rpcdump.cpp
@@ -915,0 +969,10 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+
^---- failure generated from test/lint/lint-whitespace.sh
```

MeshCollider commented:
Concept ACK

sipa commented:
I've also discovered when writing this that `importmulti` does not actually require that `watchonly` is set when no solvability is desired. I've kept the existing behavior for now, as it seems pretty invasive to people who may be relying on that, though I've added a TODO.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -840,195 +920,144 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+        bool spendable = !watchOnly;
+        if (!watchOnly) {
+            ProcessSolvingImportStep(script, import_data, SolverContext::TOP);
+            if (import_data.redeemscript) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Redeemscript provided for a non-P2SH script");
```
Why check this here, rather than passing `watchOnly` into `ProcessSolvingImportStep` and check there?

sipa replied:
If `watchonly` is true, there is nothing to do for `ProcessSolvingImportStep` (see other comment).

Sjors reviewed (commented):
Concept ACK. The new recursive `ProcessSolvingImportStep` and `SolverContext` are a lot more readable! I imagine that function lends itself to (future) unit tests better.

There's still a bunch of processing and checking happening outside of `ProcessSolvingImportStep` though, which is less easy to follow, but at least that code is now shorter.

Do the existing tests cover the new sanity checks you added?

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -805,6 +805,87 @@ UniValue dumpwallet(const JSONRPCRequest& request)
     return reply;
 }
 
+struct ImportData
```
🎉

sipa committed:
Add release notes

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -805,6 +805,87 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+};
+
+static void ProcessSolvingImportStep(const CScript& script, ImportData& data, SolverContext ctx)
+{
```
Done.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -805,6 +805,87 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+static void ProcessSolvingImportStep(const CScript& script, ImportData& data, SolverContext ctx)
+{
+    std::vector<std::vector<unsigned char>> solverdata;
+    txnouttype typ = Solver(script, solverdata);
```
Done.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -840,195 +920,144 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-
             std::vector<unsigned char> vData(ParseHex(output));
             script = CScript(vData.begin(), vData.end());
             if (!ExtractDestination(script, dest) && !internal) {
```
I think that would be completely the wrong place; perhaps the control flow isn't clear?

`ProcessSolvingImportStep` is a recursive function that matches up the parsed solvability information (only the `keys`, `pubkeys`, `redeemscript`, and `witnessscript` fields) with the provided address/script, recursing into subscripts as needed. Anything that isn't specific to the type of script being imported doesn't belong there. In particular, labels are unrelated to script types.

Sjors replied:
The rename and comments help. Still `ProcessImport` is fairly large with lots of `if` branches and loops, so splitting it into multiple functions would help. The first function could be `prepare_import_data`, the second `verify_and_import_data` which calls `RecurseImportData` and then a third `add_to_wallet`. Not sure if they can be cleanly disentangled. Maybe some other time.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -840,195 +920,144 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-            if (!pwallet->HaveCScript(witness_id) && !pwallet->AddCScript(witness_script)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2wsh witnessScript to wallet");
-            }
+        // TODO: it seems we're not actually requiring watchonly to be set for non-solvable inputs, fix this
```
That's a great idea. I've implemented that and taken it a bit further. I think all new error conditions (that don't signify seriously broken things like P2SH-inside-P2SH) are now reported as warnings instead of causing the RPC to fail.

Perhaps we want to introduce a new mode where all these warnings become errors, though `-deprecatedrpc`, but let's discuss that in a separate RPC.

sipa commented:
> There's still a bunch of processing and checking happening outside of ProcessSolvingImportStep though, which is less easy to follow, but at least that code is now shorter.

That's intentional. It's a recursive function to deal with analysing script specific solvability information. Anything that isn't specific to the script being imported isn't in there.

> Do the existing tests cover the new sanity checks you added?

Not yet, will add those soon.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -805,6 +805,91 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+struct ImportData
+{
+    // Input data
+    std::unique_ptr<CScript> redeemscript; //! Provided redeemScript; will be mvoed to `import_scripts` if relevant.
```
nit: typo "mvoed"

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -840,198 +924,160 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
 
         UniValue result = UniValue(UniValue::VOBJ);
         result.pushKV("success", UniValue(true));
+        if (warnings.size()) result.pushKV("warnings", warnings);
```
Probably needs an RPC example:
<img width="909" alt="schermafbeelding 2018-10-26 om 13 30 19" src="https://user-images.githubusercontent.com/10217/47546368-89931280-d923-11e8-89e6-811162e6580b.png">

MeshCollider commented:
utACK https://github.com/bitcoin/bitcoin/pull/14565/commits/33855fec46dbc3a7cad76875ca2a7660fcc19e92

instagibbs reviewed (commented):
looking pretty good so far

instagibbs commented:
I noticed we have no `ischange` tests yet, and would be a good time to add those.

instagibbs commented:
utACK 33855fec46dbc3a7cad76875ca2a7660fcc19e92 aside from the `ischange` test, I can't seem to convince myself we even honor the flag...

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,6 +804,91 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    std::map<CKeyID, CKey> privkeys;
+
+    // Output data
+    std::set<CScript> import_scripts;
```
Instead of a `std::set`, could `std::vector` be used for these in order to preserve the order in which things will be imported? This is useful for #14075 where we want to have the order in which things are added to the wallet be the order that was specified in the import.

instagibbs replied:
A few more changes would be required, see a fairly minimal change set(I made vectors of all the sets): https://gist.github.com/instagibbs/bd11e8666d929d84e4abc2214962e752

achow101 replied:
This has been implemented in #14075 as that is where this behavior is desired.

Sjors replied:
I think it's always desired, particularly when someone imports private keys, as those would get added to the keypool in order of their occurrence in the wallet IIUC.

sipa replied:
@Sjors As of this PR, there is no reason for doing so as we can't import into the keypool yet. Follow-ups can add the ordering.

Sjors commented:
@achow101 wrote in inline comment:

> Instead of a `std::set`, could `std::vector` be used for these in order to preserve the order in which things will be imported? This is useful for #14075 where we want to have the order in which things are added to the wallet be the order that was specified in the import.

It's also better for recoverability if the wallet strives to use derivation paths in ascending order. Otherwise the user needs to remember what range they imported before.

It would be also keep us a bit closer to honouring a BIP44-style gap limit (if the user generates more than 20 addresses without receiving coins on any of them, we'd still break that gap, but at least that's not the common case).

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +923,160 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-        if (::IsMine(*pwallet, scriptpubkey_script) == ISMINE_SPENDABLE) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+        // Watchonly and private keys
+        if (watchOnly && import_data.privkeys.size()) {
```
This one is an error, as it's always been there. The other one is just a warning.

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,165 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+            }
+        }
+        if (!have_solving_data || !::IsMine(*pwallet, script)) { // Always call AddWatchOnly for watchOnly, so that watch timestamp gets updated
```
I believe the `IsMine` check here means "related" scripts won't get added to the watchonly store after the first one, such as `wpkh` after `pkh`. Is this intended?

instagibbs replied:
Seems old code has no ISMINE checks aside from spendable fwiw

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -822,13 +911,12 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
         const UniValue& keys = data.exists("keys") ? data["keys"].get_array() : UniValue();
         const bool internal = data.exists("internal") ? data["internal"].get_bool() : false;
-        const bool watchOnly = data.exists("watchonly") ? data["watchonly"].get_bool() : false;
+        bool watchOnly = data.exists("watchonly") ? data["watchonly"].get_bool() : false;
```
Looks like this is never really used anymore, aside from causing a rejection of input when you also include privkeys,

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,165 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            }
+            if (!have_all_privkeys && !watchOnly) {
+                warnings.push_back("Assuming watchonly as not all private keys are provided.");
+                watchOnly = true;
```
what does changing this do exactly? this variable is never used later.

sipa replied:
Fixed.

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,165 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+            }
+        }
+        if (!have_solving_data || !::IsMine(*pwallet, script)) { // Always call AddWatchOnly for watchOnly, so that watch timestamp gets updated
```
old reference to `watchOnly` here

sipa replied:
Fixed (also made a change that sets have_solving_data to `false` when the solvability data is bad).

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,165 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Key does not match address destination");
+            // Verify whether the watchonly option corresponds to the availability of private keys.
+            bool have_all_privkeys = true;
+            for (const auto& used_key : import_data.used_keys) {
```
Done.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,159 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-                CScript scriptRawPubKey = GetScriptForRawPubKey(pubkey);
+            // Verify whether the watchonly option corresponds to the availability of private keys.
+            if (!watchOnly && std::any_of(import_data.used_keys.begin(), import_data.used_keys.end(), [&](const CKeyID& used_key){ return import_data.privkeys.count(used_key) == 0; })) {
+                warnings.push_back("Assuming watchonly as not all private keys are provided.");
```
This will still import the private keys provided, so I'm not 100% sure this message is clear, it may imply an all-or-nothingness with watchonly?

sipa replied:
Any suggestion for an improved message?

sipa replied:
Done.

meshcollider reviewed (commented):
re-utACK https://github.com/bitcoin/bitcoin/pull/14565/commits/7de84ca19a9719332879fd42569161b55e862391

Changes are some trivial variable renames/comment changes, replacement of the `default` case in the switch, changing some exceptions to warnings, and reworking the code to use `!have_solvable_data` instead of `watchOnly` during the import with a few related logic changes.

instagibbs commented:
LGTM, awaiting tests

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,159 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+        for (const auto& entry : import_data.pubkeys) {
+            const CPubKey& pubkey = entry.second;
+            const CKeyID& id = entry.first;
+            pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
```
This seems like a change in behavior. It looks mapKeyMetadata was only set for previously for private keys, not public keys.

Can you drop this line or add a comment about what effect it has? It seems like whatever effect it has might be temporary because AddWatchOnly() won't write this value to the database.

sipa replied:
I think this was a copy-paste error, mimicking the private key code where it was necessary. Fixed.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -839,198 +927,159 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
 
         UniValue result = UniValue(UniValue::VOBJ);
         result.pushKV("success", UniValue(true));
+        if (warnings.size()) result.pushKV("warnings", warnings);
```
It might be nice to return warnings alongside errors in the exceptional cases below. You could do this by moving the `result` and `warnings` declarations above the try block, and moving `result.pushKV("warnings")` and `return result` below.

sipa replied:
Good idea, done.

ryanofsky reviewed (approved):
utACK 7de84ca19a9719332879fd42569161b55e862391. This is much better organized than the current code, though I agree with Sjors and think it would be nice to break up the process import function more: https://github.com/bitcoin/bitcoin/pull/14565#discussion_r228410472

re: https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-432905054:

> I've also discovered when writing this that importmulti does not actually require that watchonly is set when no solvability is desired. I've kept the existing behavior for now, as it seems pretty invasive to people who may be relying on that, though I've added a TODO.

I don't see the TODO. Would the TODO just be to turn the "Assuming watchonly as not all private keys are provided." warning into an error?

re: https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-437238010:

> I realized I misread what "watchonly" was supposed to mean; it means it's fine if the resulting address is not spendable (I was under the assumption it meant being fine if it's not solvable).

Was having trouble understanding this. I think it is saying an earlier version of this PR allowed importing non-solvable scriptPubKeys as watch only, but now even watch-only scripts must be solvable. But is this actually a good requirement to have? Or is it just this way to avoid changing behavior?

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,6 +804,95 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    WITNESS_V0, //! P2WSH witnessScript
+};
+
+// Analyse the provided script, process the import data accordinging, and recurse into subscripts if necessary.
```
Fixed, done.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,6 +804,95 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        auto subscript = std::move(import_data.witnessscript);
+        if (!subscript) return "missing witnessscript";
+        if (CScriptID(*subscript) != id) return "witnessScript does not match the scriptPubKey or redeemScript";
+        import_data.import_scripts.emplace(script); // Special rule for IsMine: native P2WSH requires the TOP script imported
```
There is a comment in script/ismine.cpp about this. It was necessary to prevent an attack before SegWit activation (where someone who owes you would be able to take (another) pubkey they knew was yours, construct a P2WPKH address for it, and pay you; the wallet would show it as an incoming payment, despite it being spendable by anyone before activation). It was a terrible hack, but a necessary one, and actually what motivated me to start working on the descriptors idea (which would let us make explicit what we treat as ours, rather than relying on what we can spend).

sipa replied:
Done.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,6 +804,95 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        if (!subscript) return "missing redeemscript";
+        if (CScriptID(*subscript) != id) return "redeemScript does not match the scriptPubKey";
+        import_data.import_scripts.emplace(*subscript);
+        RecurseImportData(*subscript, import_data, ScriptContext::P2SH);
```
Good point. Fixed, and turned it all into returns.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -804,6 +804,95 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    // Input data
+    std::unique_ptr<CScript> redeemscript; //! Provided redeemScript; will be moved to `import_scripts` if relevant.
+    std::unique_ptr<CScript> witnessscript; //! Provided witnessScript; will be moved to `import_scripts` if relevant.
+    std::map<CKeyID, CPubKey> pubkeys;
```
Oh, I wasn't even away they weren't being used (they were in an earlier version of this code). Moved them up to local variables.

sipa commented:
> re: #14565 (comment):
>> I've also discovered when writing this that importmulti does not actually require that watchonly is set when no solvability is desired. I've kept the existing behavior for now, as it seems pretty invasive to people who may be relying on that, though I've added a TODO.
> I don't see the TODO. Would the TODO just be to turn the "Assuming watchonly as not all private keys are provided." warning into an error?

It refers to an earlier misunderstanding I had, which I first tried to fix the even-more-wrong way. I've updated the comment now.

> re: #14565 (comment):
>> I realized I misread what "watchonly" was supposed to mean; it means it's fine if the resulting address is not spendable (I was under the assumption it meant being fine if it's not solvable).
> Was having trouble understanding this. I think it is saying an earlier version of this PR allowed importing non-solvable scriptPubKeys as watch only, but now even watch-only scripts must be solvable. But is this actually a good requirement to have? Or is it just this way to avoid changing behavior?

Sorry for the confusion. Your comment made me read through history and make a few more changes. Here is my understanding:
* The initial *intended* behavior (see https://github.com/bitcoin/bitcoin/pull/7551#issuecomment-214397377) of "watchonly" was that it would let the user indicate explicitly that he did not desire spendability (i.e. the RPC should import the address/script as watched even though not all private keys were available).
* In #7551 his was not implemented correctly; instead of making sure that the result was spendable when "watchonly" was *not* specified, it would refuse importing private keys entirely when it *was* specified. I don't think that behavior made any sense, but embarrasingly, I don't think anyone ever noticed.
* When I first opened this PR, I misunderstood the intent and implementation, and assumed "watchonly" was intended to (but still failed to) permit non-*solvable* imports rather than non-*spendable* ones (that's what https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-432905054 refers to). As a result, I made a number of changes initially to correct this.
* Later (https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-437238010) I discovered that the intent was permitting non-spendability rather than non-solvability (which makes a lot more sense; it required fewer code and test changes, and importing something non-solvable isn't prone to error, while something non-spendable is it - you may randomly miss a private key), and adapted the PR to work towards.
* Now with your comment above, I noticed the earlier "refuse private keys when watchonly is set" behaviour, and reverted it.

Overall, as errors/warnings go, I think that the result of the above is that a number of previously permitted (but unnecessary) ways of calling this RPC now get detected. As it has a reasonable chance of breaking usage, I'm turning as many of those as possible into warnings into failures. All previously supported things should still work (but some more).

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,162 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                for (const auto& privkey : privkeys) {
+                    if (import_data.used_keys.count(privkey.first) == 0) {
+                        warnings.push_back("Ignoring irrelevant private key.");
+                        break;
```
The warning says that the irrelevant private key will be ignored, but it looks like the private key will still be imported anyways.

Also, why exit this loop when an irrelevant private key is encountered?

sipa replied:
Nice catch, fixed.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,162 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+                }
+                for (const auto& pubkey : pubkeys) {
+                    if (import_data.require_keys.count(pubkey.first) == 0) {
+                        warnings.push_back("Ignoring public key as it doesn't appear inside P2PKH or P2WPKH.");
```
Same thing as with the private keys, the pubkey will still be imported even when it is considered irrelevant.

sipa replied:
Fixed.

achow101 commented:
utACK 8a079e88f7a00cfbef0bb8b1d2016e5157343091

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -819,9 +819,97 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    // Output data
+    std::set<CScript> import_scripts;
+    std::set<CKeyID> used_keys; // Import these private keys if available
+    std::set<CKeyID> require_keys; // These public keys are required for solvability
```
I think it'd be good to mention that `require_keys` is a subset of `used_keys`. Could also choose a more normalized representation like `std::map<CKeyID, bool /* required */> used_keys`.

sipa replied:
Done.

ryanofsky reviewed (approved):
utACK 8a079e88f7a00cfbef0bb8b1d2016e5157343091, but I would make two suggestions:

- It would be good to update the importmulti "watchonly" documentation. I think the "only allowed if keys are empty" part is no longer strictly true. Also I think the documentation would be clearer if it just said that "watchonly" has no actual effect on the import, and only disables a warning if there are missing private keys.
- It would be good to update the release notes to say if things that were previously errors are now warnings.

Changes in this PR since my last review:

- Dropped the "Watch-only addresses should not include private keys" error. It looks like an "Ignoring irrelevant private key" warning will now trigger instead.
- Dropped line that stores `nCreateTime` for watch only keys in the wrong map (`mapKeyMetadata` instead of `m_script_metadata`).
- Now keeps warnings if there is an exception.
- Rebased due to conflict with #14679
- Removed `ImportData` struct  `pubkeys` and `privkeys` members, replacing with local variables.
- Added `RecurseImportData` function comment and replacing breaks with returns.

re: https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-440105516

> I discovered that the intent was permitting non-spendability rather than non-solvability (which makes a lot more sense; it required fewer code and test changes, and importing something non-solvable isn't prone to error, while something non-spendable is - you may randomly miss a private key)

I still don't understand this. There are three cases, right?

- Solvable and spendable
- Solvable but not spendable
- Not solvable, not spendable

It seems like all three cases are permitted in this PR, except there will be warnings if unused keys are provided, and errors if solving data is provided but some keys are missing. I guess I don't know if this is the final state or you want to make more changes in the future.

sipa commented:
> I still don't understand this. There are three cases, right?
>
> * Solvable and spendable
> * Solvable but not spendable
> * Not solvable, not spendable
>
> It seems like all three cases are permitted in this PR, except there will be warnings if unused keys are provided, and errors if solving data is provided but some keys are missing. I guess I don't know if this is the final state or you want to make more changes in the future.

Right, that's correct. The way to use them is:

* Solvable and spendable: give all scripts, all private keys
* Solvable but not spendable: give all scripts, all pubkeys for the pubkeyhashes in those scripts, plus optionally some (but not all) private keys, plus "watchonly".
* Not solvable, not spendable: give no scripts or keys, plus "watchonly".

The reason for the "watchonly" is distinguishing between solvable/spendable and solvable/unspendable: even if the non-spendable case you may want to import some of the related private keys. If you accidentally miss one of the keys however, the result will not be spendable. To make sure this is intentional, a warning will be given, which can be suppressed by saying "watchonly".

Sjors commented:
> If you accidentally miss one of the keys however, the result will not be spendable. To make sure this is intentional, a warning will be given, which can be suppressed by saying "watchonly".

Now that's a sentence worthy of going into the release notes (or help).

sipa commented:
This PR absolutely needs more tests, but I won't have time for that the next few days at least. If someone is interested in helping out, I think it would be useful to rewrite the existing tests to use the following pattern:K
* Keys are generated on node1
* Addresses/scripts/pubkeys are computed from those keys
* An `importmulti` query is constructed
* A function is called which is passed all this information, which calls `getaddressinfo` on the script/address itself (if it exists), but also calls it for all subscripts, P2SH/P2WSH wrapped versions, P2PK and P2PKH and P2WPKH versions of the involved keys, and tests that only the desired things are spendable/solvable/watched.

jnewbery reviewed (commented):
This is a nice improvement. I've reviewed and left a bunch of nits, mostly around making error messages clearer.

I think this comment: https://github.com/bitcoin/bitcoin/pull/14565#issuecomment-442730298 should go in release notes or docs.

> This PR absolutely needs more tests, but I won't have time for that the next few days at least. If someone is interested in helping out, I think it would be useful to rewrite the existing tests to use the following pattern

I've started refactoring the `wallet_importmulti.py` into this style. It's a WIP here: https://github.com/jnewbery/bitcoin/tree/importmulti_tests . I hope to continue working on it this week.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
 
-        // Watchonly and private keys
-        if (watchOnly && keys.size()) {
-            throw JSONRPCError(RPC_INVALID_PARAMETER, "Watch-only addresses should not include private keys");
```
Why was this check removed? Now it looks like if you provide a private key but set watchonly to true, it will still import that private key?

sipa replied:
Yes, how else would you import a private key for something you don't have all keys for? The old behavior was bogus.

MeshCollider replied:
Ah, so now the only point of the watchonly parameter is to be explicit so we don't get a warning thrown

jnewbery replied:
I think the RPC help text for `watchonly` needs to be updated. It's currently `Stating whether matching outputs should be considered watched even when they're not spendable, only allowed if keys are empty`.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/14565#discussion_r239573986

> I think the RPC help text for watchonly needs to be updated.

I suggested one possible update in https://github.com/bitcoin/bitcoin/pull/14565#pullrequestreview-179073149:

> I think the documentation would be clearer if it just said that "watchonly" has no actual effect on the import, and only disables a warning if there are missing private keys.

sipa replied:
I'd rather not say it's just disabling a warning, as it may change to a failure later.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            CPubKey pubkey = key.GetPubKey();
+            const CKeyID& id = entry.first;
+            assert(key.VerifyPubKey(pubkey));
+            pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
```
> Is there a good reason that the key's createtime metadata is updated before adding the key to the wallet?

It would be good to add a comment, but the map needs to be updated before calling AddKeyPubKey, because AddKeyPubKey uses it here: https://github.com/bitcoin/bitcoin/blob/86ff0413bb8f8173d3b3a1987875ff40b1094926/src/wallet/wallet.cpp#L291

A good cleanup would be to drop this line and instead pass the time as an argument to AddKeyPubKey, the same way it is passed as an argument to AddWatchOnly below.

Empact commented on src/wallet/rpcdump.cpp:
```diff
@@ -819,9 +819,97 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+{
+    // Input data
+    std::unique_ptr<CScript> redeemscript; //! Provided redeemScript; will be moved to `import_scripts` if relevant.
+    std::unique_ptr<CScript> witnessscript; //! Provided witnessScript; will be moved to `import_scripts` if relevant.
```
For members, doxygen uses `//!<`
http://www.doxygen.nl/manual/docblocks.html#memberdoc

jnewbery commented:
> I've started refactoring the wallet_importmulti.py into this style. It's a WIP here: https://github.com/jnewbery/bitcoin/tree/importmulti_tests . I hope to continue working on it this week.

This work is (mostly) complete. Intermediate commits could probably do with some tidy-up, which I intend to do later today.

PR here: #14886

jnewbery commented:
> I've started refactoring the wallet_importmulti.py into this style. It's a WIP here: https://github.com/jnewbery/bitcoin/tree/importmulti_tests . I hope to continue working on it this week.

This work is (mostly) complete. Intermediate commits could probably do with some tidy-up, which I intend to do later today.

PR here: #14886

jnewbery commented:
I've started writing tests for this PR: https://github.com/jnewbery/bitcoin/tree/pr14565.tests

It's certainly not complete, but it makes a start by:
- testing the return object from `importmulti` for the right `warnings` return value.
- calling `getaddressinfo` for the different address variants when privkeys are imported.

To continue, we should:

- restructure the test so in order, it's testing:
    - P2PKH (scriptPubKey and address)
    - P2WPKH (scriptPubKey and address)
    - P2SH-P2WPKH (scriptPubKey and address)
    - bare multisig (scriptPubKey only)
    - P2SH (scriptPubKey and address)
    - P2WSH (scriptPubKey and address)
    - P2SH-P2WSH (scriptPubKey and address)
- increase coverage, so that all warnings and errors are hit.

I'm unlikely to be able to spend time on the tests for the next few days. Anyone is free to take the branch and continue building on it.

sipa commented on test/functional/wallet_importmulti.py:
```diff
@@ -207,25 +207,6 @@ def run_test (self):
         assert_equal(result[0]['error']['message'], 'The wallet already contains the private key for this address or script')
 
-        # Address + Private key + watchonly
-        self.log.info("Should not import an address with private key and with watchonly")
```
Yes, the tests need updating (which I think others are working on now, I'm not entirely up to date).

sipa replied:
Ok, reverted this change.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
-            // Check the redeemScript is valid
+        // Parse all arguments
+        ImportData import_data;
+        std::map<CKeyID, CPubKey> pubkeys;
```
Done. Same with `privkey_map` for consistency.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+        for (size_t i = 0; i < pubKeys.size(); ++i) {
+            const auto& str = pubKeys[i].get_str();
+            if (!IsHex(str)) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Pubkey must be a hex string");
```
Done. I thought about doing this before but worried it would complicate writing tests. Since you've volunteered to help out with that, I don't assume that'll be a problem :p

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -819,9 +819,97 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        auto subscript = std::move(import_data.witnessscript);
+        if (!subscript) return "missing witnessscript";
+        if (CScriptID(*subscript) != id) return "witnessScript does not match the scriptPubKey or redeemScript";
+        import_data.import_scripts.emplace(script); // Special rule for IsMine: native P2WSH requires the TOP script imported
```
Good idea, done.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -854,211 +941,167 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            CPubKey pubkey = key.GetPubKey();
+            CKeyID id = pubkey.GetID();
+            if (pubkeys.count(id)) {
+                warnings.push_back("Both public and private key provided.");
```
I realized that there is very little reason for forbidding this (or even warning about it). I just removed the warning.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -819,9 +819,97 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    case TX_NULL_DATA:
+        return "unspendable script";
+    }
+    return "";
```
I wish there was a way to tell the compiler "Generate a warning if you can't prove this code is unreachable", but an assert in this case seems not worth the risk.

sipa commented:
I've addressed many of the comments above, rebased on top of the new test changes from #14886, and made a few behavior changes as well:
* Instead of sticking to a "what used to be an error and is still not supported remains an error, other things are warnings", as that resulted in pretty inconsistent behavior. I think I've changed things to be more consistently only errorring for actual failures. I still think we should do a deprecatedrpc thing to turn all warnings into errors, preferably in the same release, but maybe not in this PR.
* Added a warning for specifying "watchonly" while providing all private keys.
* Removed the special case error for redeemscript-for-P2WSH, while all other mismatches of this type are just warnings that result in non-solvability.
* More explanations in warning messages.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -860,222 +949,177 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            const CKeyID& id = entry.first;
+            assert(key.VerifyPubKey(pubkey));
+            pwallet->mapKeyMetadata[id].nCreateTime = timestamp;
+            if (!pwallet->HaveKey(id) && !pwallet->AddKeyPubKey(key, pubkey)) {
```
Suggested comment (above):
> // If the private key is not present in the wallet, insert it or add it to an existing public key.

sipa replied:
There isn't such a thing as "adding it to an existing public key", private and public keys are pretty much unrelated in every way in the wallet (the former is needed for signing and affects spendability; the latter is needed just to recurse into PKH/WPKH and not anything else). I've added some of the test you suggest, but dropped the pubkey part.

Sjors reviewed (approved):
utACK d6f2b34, but I suggest waiting for #14491 (descriptor support) rebase, as an extra check that we didn't miss anything here.

> deprecatedrpc thing to turn all warnings into errors

Or just add that as an option? It seems like a useful feature, especially since we can't delete keys and don't have a dry-run option. Different PR is fine.

promag commented on doc/release-notes-14565.md:
```diff
@@ -0,0 +1,5 @@
+Low-level RPC changes
+---------------------
+
+The `importmulti` RPC will now contain a new `warnings` field with strings
```
nit, could say:
 - it's per request result
 - it's not present if there are no warnings

sipa replied:
Done.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -836,9 +836,98 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+
+// Analyse the provided scriptPubKey, determining which keys and which redeem scripts from the ImportData struct are needed to spend it, and mark them as used.
+// Returns an error string, or the empty string for success.
+static std::string RecurseImportData(const CScript& script, ImportData& import_data, const ScriptContext script_ctx)
```
IMO could do better than empty string as success — despite the fact that it works.

Alternatives:
- could return `std::pair<bool, std::string>`
- could have return `bool` and add argument `std::string& error`

promag reviewed (commented):
PR description is misleading, especially:

> It makes sure no changes to the wallet are made when an error in the input exists.

should state that it's a per request check.

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -836,9 +836,98 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+    }
+    case TX_PUBKEYHASH: {
+        CKeyID id = CKeyID(uint160(solverdata[0]));
+        import_data.used_keys[id] = true;
```
`emplace()` doesn't overwrite the entry if it already exists (which could possible leave it with required=false).

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -836,9 +836,98 @@ UniValue dumpwallet(const JSONRPCRequest& request)
+        if (script_ctx == ScriptContext::WITNESS_V0) throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Trying to nest P2SH inside a P2WSH");
+        assert(script_ctx == ScriptContext::TOP);
+        CScriptID id = CScriptID(uint160(solverdata[0]));
+        auto subscript = std::move(import_data.redeemscript);
```
Done.

sipa commented:
@Sjors 

> Or just add that as an option? It seems like a useful feature, especially since we can't delete keys and don't have a dry-run option. Different PR is fine.

I'd really like to avoid that. There is no reason for that to be an option (which we may need to maintain forever), as the intended behavior is that these things are just errors, always. The only reason why they aren't is because we don't want to break existing software immediately.

MeshCollider commented on src/wallet/rpcdump.cpp:
```diff
@@ -860,222 +949,178 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
         }
+        for (const auto& entry : pubkey_map) {
+            const CPubKey& pubkey = entry.second;
+            const CKeyID& id = entry.first;
```
Do we also need to update `mapKeyMetadata` here to include the timestamp?

Sjors commented:
re-utACK d6f2b34

practicalswift commented on test/functional/wallet_importmulti.py:
```diff
@@ -119,9 +119,13 @@ def get_multisig(self):
         result = self.nodes[1].importmulti([req])
+        observed_warnings = []
+        if 'warnings' in result[0]:
+           observed_warnings = result[0]['warnings']
```
Indentation is not a multiple of four :-)

practicalswift commented on test/functional/wallet_importmulti.py:
```diff
@@ -119,9 +119,13 @@ def get_multisig(self):
                         script_to_p2sh_p2wsh(script_code))  # p2sh-p2wsh addr
 
-    def test_importmulti(self, req, success, error_code=None, error_message=None):
+    def test_importmulti(self, req, success, error_code=None, error_message=None, warnings=[]):
```
I suggest using `warnings=None` instead to clarify that `warnings` is not meant to be remembered across calls.

Background:

```
>>> def test(i, i_arr=[]):
...     i_arr.append(i)
...     return i_arr
...
>>> test(1)
[1]
>>> test(2)
[1, 2]
>>> test(3)
[1, 2, 3]
```

Suggested alternative:

```
>>> def test(i, i_arr=None):
...     if i_arr is None:
...         i_arr=[]
...     i_arr.append(i)
...     return i_arr
...
>>> test(1)
[1]
>>> test(2)
[2]
>>> test(3)
[3]
```

Sjors replied:
Yikes, I thought only Javascript behaved like that...

jnewbery reviewed (commented):
A bunch of comments, mostly around error messages. Please treat these all as nits. They shouldn't hold up merge and could be addressed in a follow-up PR.

This is certainly an improvement.

I've expanded the release notes here: https://github.com/jnewbery/bitcoin/commit/0348531309f64e669884f31b4de64efe0888bdd7. Feel free to squash into your commit if you think it's an improvement.

utACK eacff95de4751b500f1cef623e4024918dcb05bb

sipa commented on src/wallet/rpcdump.cpp:
```diff
@@ -860,222 +949,178 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
+            const auto& str = keys[i].get_str();
+            CKey key = DecodeSecret(str);
+            if (!key.IsValid()) {
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
```
I really don't think that putting private keys in error messages is a good idea. They may get logged unintentionally, etc.

promag replied:
Absolutely agree with @sipa. Could show a couple chars only or the error message could say the index of the invalid entry?

jnewbery replied:
Yes, very good point. I think Promag's suggestions are good.

In any case, this doesn't need to be included in this PR. A future PR to improve logging and test all failure modes would be nice.

ryanofsky reviewed (approved):
utACK eacff95de4751b500f1cef623e4024918dcb05bb. Changes since last review: rebasing, warning about unused redeem scripts instead of erroring, changing "Key does not match address destination" error to "some required keys are missing", dropping warning about private keys accompanied by public keys, adding warning if watchonly is passed along with private keys, updating RPC documentation, adding key information to errors and warnings, changing internal `used_keys` representation, and  renaming / moving some local variables.
